script/deployment/root/DeployCheckpointManager.s.sol

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.19;
+
+import "forge-std/Script.sol";
+
+import {CheckpointManager} from "contracts/root/CheckpointManager.sol";
+import {TransparentUpgradeableProxy} from "@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol";
+
+import {IBLS} from "contracts/interfaces/common/IBLS.sol";
+import {IBN256G2} from "contracts/interfaces/common/IBN256G2.sol";
+import {ICheckpointManager} from "contracts/interfaces/root/ICheckpointManager.sol";
+
+abstract contract CheckpointManagerDeployer is Script {
+    function deployCheckpointManager(
+        address proxyAdmin,
+        IBLS newBls,
+        IBN256G2 newBn256G2,
+        uint256 chainId_,
+        ICheckpointManager.Validator[] memory newValidatorSet
+    ) internal returns (address logicAddr, address proxyAddr) {
+        bytes memory initData = abi.encodeCall(
+            CheckpointManager.initialize,
+            (newBls, newBn256G2, chainId_, newValidatorSet)
+        );
+
+        vm.startBroadcast();
+
+        CheckpointManager checkpointManager = new CheckpointManager();
+
+        TransparentUpgradeableProxy proxy = new TransparentUpgradeableProxy(
+            address(checkpointManager),
+            proxyAdmin,
```

Comment:
Here the proxy admin is passed in, I recommend deploying a proxyAdmin contract within the deployment script. If an EOA, multisig, timelock, etc. is passed as the proxyAdmin by accident, this means that this account won't be able to call any function within the contract. So if the proxy admin is also an owner on the deployed contract, it couldn't call any of the protected functions.
More info here:
https://blog.openzeppelin.com/the-transparent-proxy-pattern

Here is how it would look like:

~import {TransparentUpgradeableProxy} from "@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol";~
```solidity
import {TransparentUpgradeableProxy, ProxyAdmin} from "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
```
```solidity
ProxyAdmin admin = new ProxyAdmin();
admin.transferOwnership(proxyAdmin);

TransparentUpgradeableProxy proxy = new TransparentUpgradeableProxy(
  address(implementation),
  address(admin),
  initData
);
```

I would recommend to deploy a single proxy admin contract that can upgrade all contracts. This would reduce complexity as one wouldn't need to find the proxy admin to upgrade the correct contract, the central proxy admin contract would be able to upgrade any contract.

Reply:
> This would reduce complexity as one wouldn't need to find the proxy admin to upgrade the correct contract

I like this point. Thanks.